File: backend/app/models/post.rb
# frozen_string_literal: true

# 運営からユーザーへ届ける読みもの（お知らせ・更新情報・コラム）。
class Post < ApplicationRecord
  CATEGORIES = %w[news update column].freeze
  CATEGORY_LABELS = {
    "news" => "お知らせ",
    "update" => "更新情報",
    "column" => "コラム"
  }.freeze

  # 本文に使える塊の種類。フロント側の描画と合わせること
  BLOCK_TYPES = %w[p h2 ul quote].freeze

  TITLE_MAX_LENGTH = 200
  SLUG_FORMAT = /\A[a-z0-9][a-z0-9-]*\z/

  belongs_to :author, class_name: "User", optional: true

  # 見出し画像。一覧の行頭と、記事の先頭に出す。
  # サムネを別に持つのは、一覧で原寸を並べると読み込みが重くなるため（ボックスと同じ作り）
  has_one_attached :cover_image
  has_one_attached :cover_thumb

  validates :slug, presence: true, uniqueness: true, format: { with: SLUG_FORMAT }
  validates :category, inclusion: { in: CATEGORIES }
  validates :title, presence: true, length: { maximum: TITLE_MAX_LENGTH }
  validate :body_must_be_blocks

  scope :published, -> { where.not(published_at: nil).where(published_at: ..Time.current) }
  scope :in_category, ->(category) { CATEGORIES.include?(category.to_s) ? where(category: category) : all }
  # 留めたものを先頭に、あとは新しい順
  scope :for_listing, -> { order(pinned: :desc, published_at: :desc, created_at: :desc) }

  # 記事の中に出す画像の実体。URL の組み立ては配信元を知っている側（コントローラ）に任せる。
  # 出さない設定のときは nil を返す（画面は枠ごと出さない）。
  #
  # 名前に display / list と付けるのは、`cover_thumb_blob` が
  # has_one_attached :cover_thumb の生成する関連と衝突するため（上書きすると無限再帰する）
  def cover_display_blob
    return nil unless cover_visible? && cover_image.attached?

    cover_image.blob
  end

  # 一覧に出す画像。原寸を並べると読み込みが重いので、あればサムネを使う
  def cover_list_blob
    return nil unless cover_visible?

    cover_thumb.attached? ? cover_thumb.blob : cover_display_blob
  end

  def published?
    published_at.present? && published_at <= Time.current
  end

  def delivered?
    delivered_at.present?
  end

  def category_label
    CATEGORY_LABELS[category] || category
  end

  # 平文から本文の塊を組み立てる。
  # 書く側に構造化を強いると続かないので、見出しと箇条書きだけ約束事にする。
  #   「## 」で始まる行 … 見出し
  #   「- 」で始まる行  … 箇条書き（続く行はまとめる）
  #   「> 」で始まる行  … 引用
  #   空行             … 段落の切れ目
  def self.blocks_from_text(text)
    blocks = []
    list = []

    flush_list = lambda do
      blocks << { "type" => "ul", "items" => list.dup } if list.any?
      list.clear
    end

    text.to_s.split(/\r?\n/).each do |raw|
      line = raw.strip
      if line.start_with?("- ")
        list << line.delete_prefix("- ").strip
        next
      end

      flush_list.call
      next if line.blank?

      if line.start_with?("## ")
        blocks << { "type" => "h2", "text" => line.delete_prefix("## ").strip }
      elsif line.start_with?("> ")
        blocks << { "type" => "quote", "text" => line.delete_prefix("> ").strip }
      else
        blocks << { "type" => "p", "text" => line }
      end
    end
    flush_list.call
    blocks
  end

  # 編集画面へ戻すための平文
  def body_as_text
    Array(body).filter_map do |block|
      case block["type"]
      when "h2" then "## #{block['text']}"
      when "quote" then "> #{block['text']}"
      when "ul" then Array(block["items"]).map { |item| "- #{item}" }.join("\n")
      else block["text"]
      end
    end.join("\n\n")
  end

  private

  # 知らない種類の塊は描けないので、保存の時点で弾く
  def body_must_be_blocks
    unless body.is_a?(Array)
      errors.add(:body, "の形式が正しくありません")
      return
    end

    return if body.all? { |block| block.is_a?(Hash) && BLOCK_TYPES.include?(block["type"].to_s) }

    errors.add(:body, "に扱えない種類が含まれています")
  end
end

File: frontend/src/types/post.ts
// 運営からの読みもの（お知らせ・更新情報・コラム）。
// バックエンド /api/v1/posts・/api/v1/admin/posts のレスポンスに対応。

export type PostCategory = 'news' | 'update' | 'column'

export const POST_CATEGORIES: PostCategory[] = ['news', 'update', 'column']

// お知らせの面に出す種別。コラムは読みたい理由が違うので /blog へ回す
export const NEWS_CATEGORIES: PostCategory[] = ['news', 'update']

export const POST_CATEGORY_LABELS: Record<PostCategory, string> = {
  news: 'お知らせ',
  update: '更新情報',
  column: 'コラム',
}

// 本文の塊。フロントの描画と合わせること（バックエンドの Post::BLOCK_TYPES と対応）
export type PostBlock =
  | { type: 'p'; text: string }
  | { type: 'h2'; text: string }
  | { type: 'ul'; items: string[] }
  | { type: 'quote'; text: string }

export interface Post {
  slug: string
  category: PostCategory
  category_label: string
  title: string
  excerpt: string | null
  tags: string[]
  reading_minutes: number | null
  pinned: boolean
  published_at: string | null
  /** 見出し画像。一覧はサムネ、記事の中は原寸。出さない設定なら null */
  image_url?: string | null
  /** 詳細でのみ返る */
  body?: PostBlock[]
}

// 運営画面で扱う形（下書きも含む）
export interface AdminPost {
  id: string
  slug: string
  category: PostCategory
  category_label: string
  title: string
  excerpt: string | null
  /** 編集用の平文。保存時もこの形で送る */
  body_text: string
  tags: string[]
  reading_minutes: number | null
  pinned: boolean
  published: boolean
  published_at: string | null
  /** draft（下書き）/ scheduled（予約）/ published（公開）。published だけでは予約が下書きに見える */
  status: 'draft' | 'scheduled' | 'published'
  /** 読まれた回数 */
  views_count: number
  delivered_at: string | null
  author_email: string | null
  /** 添付があっても、これが false なら一覧・記事に出さない */
  cover_visible: boolean
  image_url: string | null
  updated_at: string
}

export interface AdminPostInput {
  slug?: string
  category?: PostCategory
  title?: string
  excerpt?: string
  body_text?: string
  tags?: string[]
  reading_minutes?: number | null
  pinned?: boolean
  published?: boolean
  cover_visible?: boolean
}
